prow/plugins/approve/approve.go

```diff
@@ -0,0 +1,374 @@
+	sort.SliceStable(comments, func(i, j int) bool {
+		return comments[i].CreatedAt.Before(comments[j].CreatedAt)
+	})
+
+	approveComments := comment.FilterComments(comments, approvalCommentMatcher(botName))
+	approversHandler := approvers.NewApprovers(
+		approvers.NewOwners(
+			filenames,
+			repo,
+			int64(pr.number)))
+	approversHandler.AssociatedIssue = findAssociatedIssue(pr.body)
+	approversHandler.RequireIssue = opts.IssueRequired
+	addApprovers(&approversHandler, approveComments)
+	// Author implicitly approves their own PR if config allows it
+	if opts.ImplicitSelfApprove {
+		approversHandler.AddAuthorSelfApprover(pr.author, pr.htmlURL+"#")
+	}
+
+	for _, user := range pr.assignees {
+		approversHandler.AddAssignees(user.Login)
+	}
+
+	notificationMatcher := comment.MungerNotificationName(approvers.ApprovalNotificationName, botName)
+
+	notifications := comment.FilterComments(commentsFromIssueComments, notificationMatcher)
+	latestNotification := notifications.GetLast()
+	latestApprove := approveComments.GetLast()
+	newMessage := updateNotification(pr.org, pr.repo, latestNotification, latestApprove, approversHandler)
+	if newMessage != nil {
+		for _, notif := range notifications {
+			if err := ghc.DeleteComment(pr.org, pr.repo, notif.ID); err != nil {
+				log.WithError(err).Errorf("Failed to delete comment from %s/%s#%d, ID: %d.", pr.org, pr.repo, pr.number, notif.ID)
+			}
+		}
+		if err := ghc.CreateComment(pr.org, pr.repo, pr.number, *newMessage); err != nil {
+			log.WithError(err).Errorf("Failed to create comment on %s/%s#%d: %q.", pr.org, pr.repo, pr.number, *newMessage)
+		}
+	}
+
+	if !approversHandler.IsApproved() {
+		if hasApprovedLabel && !humanAddedApproved(ghc, log, pr.org, pr.repo, pr.number, botName) {
+			if err := ghc.RemoveLabel(pr.org, pr.repo, pr.number, approvedLabel); err != nil {
+				log.WithError(err).Errorf("Failed to remove %q label from %s/%s#%d.", approvedLabel, pr.org, pr.repo, pr.number)
+			}
+		}
+	} else {
+		//pr is fully approved
+		if !hasApprovedLabel {
+			if err := ghc.AddLabel(pr.org, pr.repo, pr.number, approvedLabel); err != nil {
+				log.WithError(err).Errorf("Failed to add %q label to %s/%s#%d.", approvedLabel, pr.org, pr.repo, pr.number)
+			}
+		}
+	}
+	return nil
+}
+
+func humanAddedApproved(ghc githubClient, log *logrus.Entry, org, repo string, number int, botName string) bool {
+	events, err := ghc.ListIssueEvents(org, repo, number)
+	if err != nil {
+		log.WithError(err).Errorf("Failed to list issue events for %s/%s#%d.", org, repo, number)
+		return false
+	}
+	approveAddedMatcher := event.And([]event.Matcher{event.AddLabel{}, event.LabelName(approvedLabel)})
+	labelEvents := event.FilterEvents(events, approveAddedMatcher)
+	lastAdded := labelEvents.GetLast()
+	if lastAdded == nil || lastAdded.Actor.Login == "" {
+		return false
+	}
+	return lastAdded.Actor.Login != botName
+}
+
+func approvalCommentMatcher(botName string) comment.Matcher {
```

Comment:
My interaction with this system was a little unpleasant. I think the idea here is powerful but seems too slick for Go. In the end the filters end up being very simple and would be a lot easier to work with in my opinion if we did not try to create a generic matcher system. Passing through args (like the bot or org name) also made them a lot less pretty after being parameterized.

Perhaps we can just have `comment.Matcher` be a `func(comment *github.Comment) bool` instead of creating structures?

Reply:
I don't think I understand the change you are suggesting.
I don't see how changing the matcher types to functions would simplify things. The functions would still need to be parameterized by the same args that the struct matchers required.
Are you suggesting that we don't use the `comment` package at all?

Reply:
Yes, my vote would be not to use it. It seems more complex than it needs to be. Parameterizing things has made those parts a lot less self-contained. Needing to read 100+ LOC to understand:

```go
func HumanActor(mungeBotName string) Matcher {
	return And([]Matcher{
		ValidAuthor{},
		Not{BotAuthor(mungeBotName)},
	})
}
```

Instead of being immediately ready to understand

```
if comment.Author != nil && comment.Author != botName {
    // .. do work
}
```

Similar to the way we just do simple `if` to match the types of events we care about at the top of every Prow plugin today.